src/modules/components/proposal/proposalActions/proposalActionsAction/proposalActionsAction.tsx

```diff
@@ -21,33 +29,48 @@ export interface IProposalActionsActionProps extends IWeb3ComponentProps {
     /**
      * Custom component for the action
      */
-    customComponent?: ProposalActionComponent;
+    CustomComponent?: ProposalActionComponent;
 }
 
 export const ProposalActionsAction: React.FC<IProposalActionsActionProps> = (props) => {
-    const { action, index, name, customComponent, ...web3Props } = props;
-
-    const ActionComponent = customComponent ?? proposalActionsUtils.getActionComponent(action);
+    const { action, index, name, CustomComponent, ...web3Props } = props;
 
     const { copy } = useOdsModulesContext();
 
+    const ActionComponent = useMemo(() => {
+        if (CustomComponent) {
+            return <CustomComponent action={action} {...web3Props} />;
+        }
+
+        if (proposalActionsUtils.isWithdrawTokenAction(action)) {
+            return <ProposalActionWithdrawToken action={action} {...web3Props} />;
+        } else if (proposalActionsUtils.isTokenMintAction(action)) {
+            return <ProposalActionTokenMint action={action} {...web3Props} />;
+        } else if (proposalActionsUtils.isUpdateMetadataAction(action)) {
+            return <ProposalActionUpdateMetadata action={action} {...web3Props} />;
+        } else if (proposalActionsUtils.isChangeMembersAction(action)) {
+            return <ProposalActionChangeMembers action={action} {...web3Props} />;
+        } else if (proposalActionsUtils.isChangeSettingsAction(action)) {
+            return <ProposalActionChangeSettings action={action} {...web3Props} />;
+        }
+
+        return null;
+    }, [action, CustomComponent, web3Props]);
```

Comment:
I think this works for now. I want to abstract this into some kind of factory pattern. But it's great we can leverage useMemo here in the function component.

Reply:
Yeah, at least we have the handling of react components on a react component now. We can probably still improve this, let me know if you have other ideas